test/parallel_api/algorithm/alg.sorting/sort_by_key_common.h

```diff
@@ -0,0 +1,289 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _SORT_BY_KEY_COMMON_H
+#define _SORT_BY_KEY_COMMON_H
+
+#include "support/test_config.h"
+
+#include _PSTL_TEST_HEADER(execution)
+#include _PSTL_TEST_HEADER(algorithm)
+
+#include "support/utils.h"
+#if TEST_DPCPP_BACKEND_PRESENT
+#   include "support/sycl_alloc_utils.h"
+#endif
+
+#include <tuple> // std::tie
+#include <random> // std::default_random_engine, uniform_real_distribution
+#include <vector> // std::vector
+#include <iterator> // std::distance
+#include <algorithm> // std::generate, std::remove_if, std::stable_sort, std::is_sorted, std::transform, std::shuffle
+#include <type_traits> // std::is_same
+#include <unordered_map> // std::unordered_map
+
+struct StableSort{};
+struct UnstableSort{};
+
+struct DefaultComparator{};
+struct GreaterComparator
+{
+    template<typename T>
+    bool operator()(const T& lhs, const T& rhs) const {
+        return lhs > rhs;
+    }
+};
+
+struct Particle
+{
+    float mass;
+    float velocity;
+    float coordinates[3];
+    using energy_type = float;
+    energy_type energy() const
+    {
+        return 0.5 * mass * velocity * velocity;
```

Comment:
I guess, a comment "kinetic energy" should be helpfull here.

Reply:
Added